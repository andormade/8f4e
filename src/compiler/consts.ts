export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
