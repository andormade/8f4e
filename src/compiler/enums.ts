export const enum NameSection {
	FUNCTION_NAME = 0x01,
	LOCAL_NAME = 0x02,
}

export const enum ImportDesc {
	MEMORY = 0x02,
}

export const enum ExportDesc {
	FUNC = 0x00,
}

export const enum Type {
	I32 = 0x7f,
	F32 = 0x7d,
}

export const enum Instruction {
	IF = 0x04,
	ELSE = 0x05,
	END = 0x0b,
	CALL = 0x10,
	LOCAL_GET = 0x20,
	I32_LOAD = 0x28,
	F32_LOAD = 0x2a,
	I32_STORE = 0x36,
	I32_CONST = 0x41,
	F32_CONST = 0x43,
	I32_EQ = 0x46,
	I32_GE_S = 0x4e,
	I32_ADD = 0x6a,
	I32_SUB = 0x6b,
	I32_MUL = 0x6c,
	I32_DIV_U = 0x6e,
	F32_STORE = 0x83,
	F32_ADD = 0x92,
	F32_SUB = 0x93,
	F32_MUL = 0x94,
	F32_DIV = 0x95,
}

export const enum Section {
	CUSTOM = 0x00,
	TYPE = 0x01,
	IMPORT = 0x02,
	FUNCTION = 0x03,
	MEMORY = 0x05,
	EXPORT = 0x07,
	CODE = 0x0a,
}
