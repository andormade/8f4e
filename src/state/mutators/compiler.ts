import { State } from '../types';

export default function compiler(state: State, events) {
	const worker = new Worker(new URL('../../../packages/worker/src/index.ts', import.meta.url));
	// @ts-ignore shared: true
	const memoryRef = new WebAssembly.Memory({ initial: 1, maximum: 1, shared: true });

	async function recompile() {
		worker.postMessage({ memoryRef, modules: state.modules, connections: state.connections });
		state.compiler.isCompiling = true;
		state.compiler.lastCompilationStart = performance.now();
	}

	async function onWorkerMessage({ data }) {
		switch (data.type) {
			case 'midiMessage':
				events.dispatch('sendMidiMessage', data.payload);
				break;
			case 'compilationDone':
				state.compiler.memoryBuffer = new Int32Array(memoryRef.buffer);
				state.compiler.outputAddressLookup = data.payload.outputAddressLookup;
				state.compiler.isCompiling = false;
				const end = performance.now() - state.compiler.lastCompilationStart;
				state.compiler.compilationTime = end.toFixed(2);

				state.modules.forEach(module => {
					if (state.moduleTypes[module.type].transformer) {
						state.moduleTypes[module.type].transformer(
							module,
							state.compiler.memoryBuffer,
							state.compiler.outputAddressLookup
						);
					}
				});
				break;
		}
	}

	worker.addEventListener('message', onWorkerMessage);
	events.on('createConnection', recompile);
	events.on('deleteConnection', recompile);
	events.on('addModule', recompile);
	events.on('deleteModule', recompile);
	events.on('init', recompile);

	return () => {
		events.off('createConnection', recompile);
		events.off('deleteConnection', recompile);
		events.off('addModule', recompile);
		events.off('deleteModule', recompile);
		events.off('init', recompile);
	};
}
