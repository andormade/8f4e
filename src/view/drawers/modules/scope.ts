import { font, scope } from '../../spriteGenerator';

const drawer = function (engine, state, id) {
	const address = state.ui.compiler.outputAddressLookup[id + 'buffer'] / 4;
	const buffer = state.ui.compiler.memoryBuffer.slice(address, address + 80);

	engine.setSpriteLookup(scope);

	for (let i = 0; i < 80; i++) {
		engine.drawSprite(1 * i + 10, 20, buffer[i], 1, 80);
	}
};

export default drawer;
