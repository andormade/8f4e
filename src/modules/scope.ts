import addDefaultInputPositions from './helpers/addDefaultInputPositions';
import addDefaultOutputPositions from './helpers/addDefaultOutputPositions';
import { ModuleGeneratorProps, ModuleType } from '../state/types';
import { MODULE_HEIGHT_S, MODULE_WIDTH_S } from './consts';

export default function scope({ vGrid, hGrid }: ModuleGeneratorProps): ModuleType {
	return {
		category: 'Inspection',
		engine: 'scope',
		height: MODULE_HEIGHT_S * hGrid,
		inputs: addDefaultInputPositions([{ id: 'in' }], vGrid, hGrid),
		name: 'Scope',
		outputs: addDefaultOutputPositions([{ id: 'out' }], vGrid, hGrid, MODULE_WIDTH_S * vGrid),
		sliders: [],
		steppers: [],
		switches: [],
		width: MODULE_WIDTH_S * vGrid,
	};
}
