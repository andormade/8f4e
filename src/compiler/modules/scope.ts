import { i32storeLocal, i32load, localGet, localSet, i32const, i32store, ifelse } from '../wasm/instructions';
import { createFunctionBody, createLocalDeclaration } from '../wasm/sections';
import { Instruction, Type } from '../wasm/enums';
import { ModuleGenerator } from './types';

const enum Memory {
	ZERO = 0x00,
	INPUT_POINTER = 0x04,
	OUTPUT = 0x08,
	BUFFER_POINTER = 0x0c,
	BUFFER_START = 0x10,
}

const enum Locals {
	INPUT,
	BUFFER_POINTER,
	__LENGTH,
}

const BUFFER_LENGTH = Int32Array.BYTES_PER_ELEMENT * 8;

const scope: ModuleGenerator = function (moduleId, offset) {
	const functionBody = createFunctionBody(
		[createLocalDeclaration(Type.I32, Locals.__LENGTH)],
		[
			// Load input value from memory, store it in a register.
			...i32const(Memory.INPUT_POINTER + offset),
			...i32load(),
			...i32load(),
			...localSet(Locals.INPUT),

			// Load buffer pointer and store it in a register.
			...i32const(Memory.BUFFER_POINTER + offset),
			...i32load(),
			...localSet(Locals.BUFFER_POINTER),

			// Store the input value in the memory where the buffer pointer points.
			...localGet(Locals.BUFFER_POINTER),
			...localGet(Locals.INPUT),
			...i32store(),

			// Increment the value of the the buffer pointer.
			...localGet(Locals.BUFFER_POINTER),
			...i32const(Int32Array.BYTES_PER_ELEMENT),
			Instruction.I32_ADD,
			...localSet(Locals.BUFFER_POINTER),

			// Prevent the buffer pointer access out of bounds memory.
			...localGet(Locals.BUFFER_POINTER),
			...i32const(Memory.BUFFER_START + BUFFER_LENGTH + offset),
			Instruction.I32_GE_U,
			...ifelse(Type.VOID, [...i32const(Memory.BUFFER_START + offset), ...localSet(Locals.BUFFER_POINTER)]),

			// Store the buffer pointer in the memory.
			...i32const(Memory.BUFFER_POINTER + offset),
			...localGet(Locals.BUFFER_POINTER),
			...i32store(),

			...i32storeLocal(Locals.INPUT, Memory.OUTPUT + offset),
		]
	);

	return {
		moduleId,
		functionBody,
		offset,
		initialMemory: [0, Memory.ZERO + offset, 0, Memory.BUFFER_START + offset, 0, 0, 0, 0, 0, 0, 0, 0],
		memoryAddresses: [
			{ address: Memory.OUTPUT + offset, id: 'out' },
			{ address: Memory.INPUT_POINTER + offset, id: 'in', isInputPointer: true },
			{ address: Memory.BUFFER_START + offset, id: 'buffer' },
		],
	};
};

export default scope;
