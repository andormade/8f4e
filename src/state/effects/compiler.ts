import { State } from '../types';
import { compilationDone, recompile } from '../mutators/compiler';

export default function compiler(state: State, events): void {
	const worker = new Worker(new URL('../../../packages/worker/src/index.ts', import.meta.url));
	// @ts-ignore shared: true
	const memoryRef = new WebAssembly.Memory({ initial: 1, maximum: 1, shared: true });

	async function onRecompile() {
		worker.postMessage({ memoryRef, modules: state.modules, connections: state.connections });
		recompile(state);
	}

	async function onWorkerMessage({ data }) {
		switch (data.type) {
			case 'midiMessage':
				events.dispatch('sendMidiMessage', data.payload);
				break;
			case 'compilationDone':
				compilationDone(state, data, memoryRef);
				state.modules.forEach(({ type }, index) => {
					if (state.moduleTypes[type].insertState) {
						state.moduleTypes[type].insertState(
							state.modules[index].state,
							state.compiler.memoryBuffer,
							state.compiler.memoryAddressLookup[state.modules[index].id]
						);
					}
				});
				break;
		}
	}

	worker.addEventListener('message', onWorkerMessage);
	events.on('createConnection', onRecompile);
	events.on('deleteConnection', onRecompile);
	events.on('addModule', onRecompile);
	events.on('deleteModule', onRecompile);
	events.on('init', onRecompile);
}
