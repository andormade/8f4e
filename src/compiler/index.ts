import {
	createFunctionSection,
	createTypeSection,
	createFunctionType,
	createExportSection,
	createFunctionExport,
	createCodeSection,
	createFunctionBody,
	createImportSection,
	createMemoryImport,
} from './wasm/sections';

import { call, i32store } from './wasm/instructions';
import { generateOutputAddressLookup } from './initializeMemory';
import * as moduleCompilers from './modules';
import { Module } from './modules/types';
import { i32abs, i32modulo } from './wasm/helpers';
import { Type } from 'wasm-bytecode-utils';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

const compileModules = function (modules): Module[] {
	let memoryAddress = 4;
	return modules
		.filter(({ type }) => moduleCompilers[type])
		.map(({ id, type }) => {
			const module = moduleCompilers[type](id, memoryAddress);
			memoryAddress += module.initialMemory.length * Int32Array.BYTES_PER_ELEMENT;
			return module;
		});
};

const generateMemoryInitiatorFunction = function (compiledModules) {
	return compiledModules
		.map(module => {
			let pointer = module.offset;
			return module.initialMemory
				.map(value => {
					const instuction = i32store(pointer, value);
					pointer += Int32Array.BYTES_PER_ELEMENT;
					return instuction;
				})
				.flat();
		})
		.flat();
};

const compile = function (modules: object[]) {
	const compiledModules = compileModules(modules);
	const functionBodies = compiledModules.map(({ functionBody }) => functionBody);
	const functionSignatures = compiledModules.map(() => 0x00);
	const cycleFunction = compiledModules.map((module, index) => call(index + 4)).flat();
	const memoryInitiatorFunction = generateMemoryInitiatorFunction(compiledModules);
	const outputAddressLookup = generateOutputAddressLookup(compiledModules);
	const helperFunctions = [i32abs(), i32modulo()];
	const helperFunctionSignatures = [0x01, 0x02];

	return {
		codeBuffer: Uint8Array.from([
			...HEADER,
			...VERSION,
			...createTypeSection([
				createFunctionType([], []),
				createFunctionType([Type.I32], [Type.I32]),
				createFunctionType([Type.I32, Type.I32], [Type.I32]),
			]),
			...createImportSection([createMemoryImport('js', 'memory', 1, 1, true)]),
			...createFunctionSection([...helperFunctionSignatures, 0x00, 0x00, ...functionSignatures]),
			...createExportSection([createFunctionExport('init', 0x02), createFunctionExport('cycle', 0x03)]),
			...createCodeSection([
				...helperFunctions,
				createFunctionBody([], memoryInitiatorFunction),
				createFunctionBody([], cycleFunction),
				...functionBodies,
			]),
		]),
		outputAddressLookup,
		compiledModules,
	};
};

export default compile;
