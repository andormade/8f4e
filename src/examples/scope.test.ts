import { createTestModule } from '@8f4e/compiler';

import scope from './scope.asm';

let testModule;

describe('functional tests', () => {
	beforeAll(async () => {
		testModule = await createTestModule(scope);
	});

	beforeEach(() => {
		testModule.reset();
	});

	test('if the generated ast matches with the snapshot', () => {
		expect(testModule.ast).toMatchSnapshot();
	});

	test('if the wat code matches with the snapshot', () => {
		expect(testModule.wat).toMatchSnapshot();
	});

	test('if the generated memory map matches with the snapshot', () => {
		expect(testModule.memoryMap).toMatchSnapshot();
	});
});
