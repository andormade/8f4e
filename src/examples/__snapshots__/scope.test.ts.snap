// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functional tests if the generated ast matches with the snapshot 1`] = `
Array [
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "scope",
      },
    ],
    "instruction": "module",
    "lineNumber": 2,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "defaultValue",
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "private",
    "lineNumber": 4,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "in",
      },
      Object {
        "type": "identifier",
        "value": "defaultValue",
      },
    ],
    "instruction": "inputPointer",
    "lineNumber": 5,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "out",
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "output",
    "lineNumber": 6,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "counter",
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "private",
    "lineNumber": 8,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "rate",
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "private",
    "lineNumber": 9,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "buffer",
      },
      Object {
        "type": "literal",
        "value": 64,
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "array",
    "lineNumber": 10,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "bufferPointer",
      },
      Object {
        "type": "identifier",
        "value": "buffer",
      },
    ],
    "instruction": "public",
    "lineNumber": 11,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_bufferPointer",
      },
    ],
    "instruction": "local",
    "lineNumber": 13,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "local",
    "lineNumber": 14,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_input",
      },
    ],
    "instruction": "local",
    "lineNumber": 15,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "in",
      },
    ],
    "instruction": "push",
    "lineNumber": 17,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_input",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 18,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&out",
      },
    ],
    "instruction": "push",
    "lineNumber": 20,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_input",
      },
    ],
    "instruction": "push",
    "lineNumber": 21,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 22,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 24,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 25,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 27,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "rate",
      },
    ],
    "instruction": "push",
    "lineNumber": 28,
  },
  Object {
    "arguments": Array [],
    "instruction": "greaterOrEqual",
    "lineNumber": 29,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "void",
      },
    ],
    "instruction": "if",
    "lineNumber": 30,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 31,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 32,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 33,
  },
  Object {
    "arguments": Array [],
    "instruction": "else",
    "lineNumber": 34,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 35,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 1,
      },
    ],
    "instruction": "push",
    "lineNumber": 36,
  },
  Object {
    "arguments": Array [],
    "instruction": "add",
    "lineNumber": 37,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 38,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 40,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 41,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 42,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 1,
      },
    ],
    "instruction": "branch",
    "lineNumber": 43,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 44,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "bufferPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 46,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_bufferPointer",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 47,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_bufferPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 49,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_input",
      },
    ],
    "instruction": "push",
    "lineNumber": 50,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 51,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_bufferPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 54,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 4,
      },
    ],
    "instruction": "push",
    "lineNumber": 55,
  },
  Object {
    "arguments": Array [],
    "instruction": "add",
    "lineNumber": 56,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_bufferPointer",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 57,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_bufferPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 60,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&buffer",
      },
    ],
    "instruction": "push",
    "lineNumber": 61,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 256,
      },
    ],
    "instruction": "push",
    "lineNumber": 62,
  },
  Object {
    "arguments": Array [],
    "instruction": "add",
    "lineNumber": 63,
  },
  Object {
    "arguments": Array [],
    "instruction": "greaterOrEqualUnsigned",
    "lineNumber": 64,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "void",
      },
    ],
    "instruction": "if",
    "lineNumber": 65,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&buffer",
      },
    ],
    "instruction": "push",
    "lineNumber": 66,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_bufferPointer",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 67,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 68,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&bufferPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 70,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_bufferPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 71,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 72,
  },
]
`;

exports[`functional tests if the generated memory map matches with the snapshot 1`] = `
Map {
  "defaultValue" => Object {
    "byteAddress": 0,
    "default": 0,
    "id": "defaultValue",
    "lineNumber": 4,
    "relativeWordAddress": 0,
    "type": 5,
    "wordSize": 1,
  },
  "in" => Object {
    "byteAddress": 4,
    "default": 0,
    "id": "in",
    "lineNumber": 5,
    "relativeWordAddress": 1,
    "type": 3,
    "wordSize": 1,
  },
  "out" => Object {
    "byteAddress": 8,
    "default": 0,
    "id": "out",
    "lineNumber": 6,
    "relativeWordAddress": 2,
    "type": 4,
    "wordSize": 1,
  },
  "counter" => Object {
    "byteAddress": 12,
    "default": 0,
    "id": "counter",
    "lineNumber": 8,
    "relativeWordAddress": 3,
    "type": 5,
    "wordSize": 1,
  },
  "rate" => Object {
    "byteAddress": 16,
    "default": 0,
    "id": "rate",
    "lineNumber": 9,
    "relativeWordAddress": 4,
    "type": 5,
    "wordSize": 1,
  },
  "buffer" => Object {
    "byteAddress": 20,
    "default": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "id": "buffer",
    "lineNumber": 10,
    "relativeWordAddress": 5,
    "type": 0,
    "wordSize": 64,
  },
  "bufferPointer" => Object {
    "byteAddress": 276,
    "default": 20,
    "id": "bufferPointer",
    "lineNumber": 11,
    "relativeWordAddress": 69,
    "type": 6,
    "wordSize": 1,
  },
}
`;

exports[`functional tests if the wat code matches with the snapshot 1`] = `
"(module
  (type (;0;) (func))
  (import \\"js\\" \\"memory\\" (memory (;0;) 1))
  (func (;0;) (type 0)
    (local i32 i32 i32)
    i32.const 4
    i32.load
    i32.load
    local.set 2
    i32.const 8
    local.get 2
    i32.store
    i32.const 12
    i32.load
    local.set 1
    local.get 1
    i32.const 16
    i32.load
    i32.ge_s
    if  ;; label = @1
      i32.const 12
      i32.const 0
      i32.store
    else
      local.get 1
      i32.const 1
      i32.add
      local.set 1
      i32.const 12
      local.get 1
      i32.store
      br 1 (;@0;)
    end
    i32.const 276
    i32.load
    local.set 0
    local.get 0
    local.get 2
    i32.store
    local.get 0
    i32.const 4
    i32.add
    local.set 0
    local.get 0
    i32.const 20
    i32.const 256
    i32.add
    i32.ge_u
    if  ;; label = @1
      i32.const 20
      local.set 0
    end
    i32.const 276
    local.get 0
    i32.store)
  (export \\"test\\" (func 0)))
"
`;
